content/post/2019-11-21-readmes.Rmarkdown

```diff
@@ -0,0 +1,416 @@
+
+#### Number of lines
+
+A first metric we'll extract is the number of lines of the README. 
+
+```{r}
+count_lines <- function(readme_lines){
+  readme_lines %>%
+    purrr::discard(. == "") %>% # emtpy lines
+    purrr::discard(stringr::str_detect(., "\\<\\!\\-\\-")) %>% # html comments
+    length()
+}
+
+readmes <- dplyr::group_by(readmes, owner, repo) %>%
+  dplyr::mutate(lines_no = count_lines(readme[[1]])) %>%
+  dplyr::ungroup()
+
+```
+
+How long are usual READMEs? Their number of lines range from `r min(readmes$lines_no)` to `r max(readmes$lines_no)` with a median of `r median(readmes$lines_no)`.
+
+```{r nolinesplot, fig.cap="Dot plot of the number of lines in READMEs"}
+library("ggplot2")
+
+ggplot(readmes) +
+  geom_dotplot(aes(lines_no), binwidth = 5) +
+  xlab("No. of lines") +
+  scale_y_continuous(NULL, breaks = NULL) +
+  hrbrthemes::theme_ipsum(base_size = 16,
+                          axis_title_size = 16) +
+  ggtitle(glue::glue("Number of lines in a sample of {nrow(readmes)} READMEs"))
+```
+
+READMEs in our sample most often don't have more than 200 lines. Now, this metric might indicate how much a potential user needs to take in and how long they need to scroll down but we shall now look into other indicators of size: the number of lines of R code, the number of words outside of code and output.
+
+#### Other size indicators
+
+To access the numbers we're after without using too many regular expressions, [we shall convert the Markdown content to XML via `commonmark` and use XPath to parse it](https://ropensci.org/technotes/2018/09/05/commonmark/).
+
+```{r xml}
+get_xml <- function(readme_lines){
+  readme_lines %>%
+      glue::glue_collapse(sep = "\n") %>%
+      commonmark::markdown_xml(normalize = TRUE,
+                               hardbreaks = TRUE) %>%
+      xml2::read_xml() %>%
+      xml2::xml_ns_strip() -> xml
+    
+    xml2::xml_replace(xml2::xml_find_all(xml, "//softbreak"),
+                      xml2::read_xml("<text>\n</text>"))
+    
+    list(xml)
+}
+
+readmes <- dplyr::group_by(readmes, owner, repo) %>%
+  dplyr::mutate(xml_readme = get_xml(readme[[1]])) %>%
+  dplyr::ungroup()
+```
+
+This is how a single README XML looks like:
+
+```{r xmlex}
+readmes$xml_readme[[1]]
+```
+
+Let's count lines of code.
+
+
+```{r loc}
+get_code_lines <- function(xml) {
+  
+  if(is.null(xml)) {
+    return(NULL)
+  }
+  
+  xml2::xml_find_all(xml, "code_block") %>%
+    purrr::keep(xml2::xml_attr(., "info") == "r") %>%
+    xml2::xml_text() %>%
+    length
+}
+
+loc <- readmes %>%
+  dplyr::group_by(repo, owner) %>%
+  dplyr::summarise(loc = get_code_lines(xml_readme[[1]]))
+
+```
+
+The number of lines of code of READMEs range from `r min(loc$loc)` (for `r sum(loc$loc == min(loc$loc))` READMEs) to `r max(loc$loc)` with a median of `r median(loc$loc)`. The README with the most lines of code is `r glue::glue("https://github.com/{loc$owner[loc$loc == max(loc$loc)]}/{loc$repo[loc$loc == max(loc$loc)]}")`.
```

Comment:
You can append `#readme` to link to the actual README.

Reply:
awesome idea